bse/driver-jack.cc

```diff
@@ -0,0 +1,905 @@
+
+  assert_return (jack_client, devices);
+
+  const char **jack_ports = jack_get_ports (jack_client, NULL, NULL, 0);
+  if (jack_ports)
+    {
+      bool have_default_device = false;
+
+      for (uint i = 0; jack_ports[i]; i++)
+	{
+	  const char *end = strchr (jack_ports[i], ':');
+	  if (end)
+	    {
+              std::string device_name (jack_ports[i], end);
+
+	      jack_port_t *jack_port = jack_port_by_name (jack_client, jack_ports[i]);
+	      if (jack_port)
+		{
+                  const char *port_type = jack_port_type (jack_port);
+                  if (strcmp (port_type, JACK_DEFAULT_AUDIO_TYPE) == 0)
+                    {
+	              DeviceDetails &details = devices[device_name];
+	              details.ports++;
+
+                      int flags = jack_port_flags (jack_port);
+                      if (flags & JackPortIsInput)
+                        {
+                          details.input_ports++;
+                          details.input_port_names.push_back (jack_ports[i]);
+                        }
+                      if (flags & JackPortIsOutput)
+                        {
+                          details.output_ports++;
+                          details.output_port_names.push_back (jack_ports[i]);
+                        }
+                      if (flags & JackPortIsPhysical)
+                        {
+                          details.physical_ports++;
+
+                          if (!have_default_device)
+                            {
+                              /* the first device that has physical ports is the default device */
+                              details.default_device = true;
+                              have_default_device = true;
+                            }
+                        }
+                      if (flags & JackPortIsTerminal)
+                        details.terminal_ports++;
+                    }
+		}
+	    }
+	}
+      free (jack_ports);
+    }
+
+  return devices;
+}
+
+static void
+list_jack_drivers (Driver::EntryVec &entries)
+{
+  map<std::string, DeviceDetails> devices;
+  jack_client_t *jack_client = connect_jack();
+  if (jack_client)
+    {
+      devices = query_jack_devices (jack_client);
+      disconnect_jack (jack_client);
+    }
+  else
+    {
+      // should we try to generate and show an error message if connecting jack failed?
+    }
+
+  for (std::map<std::string, DeviceDetails>::iterator di = devices.begin(); di != devices.end(); di++)
+    {
+      const std::string &device_name = di->first;
+      DeviceDetails &details = di->second;
+
+      /* the default device is usually the hardware device, so things should work as expected
+       * we could show try to show non-default devices as well, but this could be confusing
+       */
+      if (details.default_device)
+        {
+          Driver::Entry entry;
+          entry.devid = device_name;
+          entry.priority = Driver::JACK;
+          entries.push_back (entry);
```

Comment:
Two Driver::Entry structures must never have the same priority. That's why we have predefined priority constants for drivers, cards, devices, subdevices. Also, we need the other Entry fields to be filled to allow GUI selection. For now, it doesn't matter too much how to match name/blurb/status fields to extra info, each is just another line in my current UI prototype, but we need *something* to display devices to the user that make them humanly identifyable.

Reply:
There are never two Driver::Entry structures here. There is a loop, so you might expect it, but basically the code picks the only device which it considers to be the "default" device from the devices map. It used generate a list of devices, but I think thats more harmful than it helps, but I left the devices query and generate list code there, in case we want to do it differently.

Jack is not very descriptive. Typically we only get the client name "system" and the port names "playback_1", "playback_2",... and we know that it is a hardware device. So ok, I'm trying to give all that information to the user in my new version, but still, we for instance do *not* get a soundcard name or anything useful. There is a Jack Metadata API which could be used to attach descriptions to clients, but at least on my system, this returns nothing. So really what we have is "system", "is hardware device", "2*in 2*out", nothing really useful for the user.